File: php-symfony/src/Service/CodeGenerator.php
<?php

namespace App\Service;

use App\Repository\LinkRepository;

class CodeGenerator
{
    private const CHARSET = 'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789';
    private const LENGTH = 6;
    private const MAX_RETRIES = 3;

    public function generate(LinkRepository $repository): string
    {
        for ($i = 0; $i < self::MAX_RETRIES; $i++) {
            $code = $this->randomCode();
            if ($repository->findOneBy(['code' => $code]) === null) {
                return $code;
            }
        }

        throw new \RuntimeException('Could not generate a unique code after ' . self::MAX_RETRIES . ' attempts');
    }

    private function randomCode(): string
    {
        $code = '';
        $max = strlen(self::CHARSET) - 1;

        for ($i = 0; $i < self::LENGTH; $i++) {
            $code .= self::CHARSET[random_int(0, $max)];
        }

        return $code;
    }
}

File: java-springboot/src/main/java/com/chopchop/service/CodeGenerator.java
package com.chopchop.service;

import com.chopchop.repository.LinkRepository;
import org.springframework.stereotype.Service;

import java.security.SecureRandom;

@Service
public class CodeGenerator {

    private static final String ALPHABET = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
    private static final int CODE_LENGTH = 6;
    private static final int MAX_ATTEMPTS = 3;

    private final SecureRandom random = new SecureRandom();
    private final LinkRepository linkRepository;

    public CodeGenerator(LinkRepository linkRepository) {
        this.linkRepository = linkRepository;
    }

    public String generate() {
        for (int i = 0; i < MAX_ATTEMPTS; i++) {
            String code = randomCode();
            if (!linkRepository.existsByCode(code)) {
                return code;
            }
        }
        throw new IllegalStateException("Failed to generate a unique code after " + MAX_ATTEMPTS + " attempts");
    }

    private String randomCode() {
        StringBuilder sb = new StringBuilder(CODE_LENGTH);
        byte[] bytes = new byte[CODE_LENGTH];
        random.nextBytes(bytes);
        for (byte b : bytes) {
            sb.append(ALPHABET.charAt(Math.abs(b % ALPHABET.length())));
        }
        return sb.toString();
    }
}
